opm/simulators/linalg/bda/BILU0.hpp

```diff
@@ -0,0 +1,118 @@
+/*
+  Copyright 2019 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef BILU0_HPP
+#define BILU0_HPP
+
+#include <config.h> // CMake
```

Comment:
No, that's not going to work.  It is always an error to
```C++
#include <config.h>
```
in a header.  That include statement always needs to be in an implementation (`.cpp`) file.

Reply:
Can you explain why? For BILU0 it does not matter, it is only ever included when OpenCL was found.
But WellContributions.hpp should be able to be used in a CUDA-only environment and in an OpenCL-only environment. Would you propose a layer of inheritance?

Reply:
>> It is always an error to
>> ```C++
>> #include <config.h>
>> ```
>> in a header.
>
> Can you explain why?

It has to do with installation and packaging.  We never install/package `config.h` which means that any header that refers directly to `config.h` will fail to compile in an installed environment.  Moreover, if we were to install `config.h` you would get the configuration of the build machine that produced the packages rather than the host system that runs the binaries.  Those could very well be different.  Finally, you will notice that there is a separate `config.h` in the top-level build directory of each module.  Those files are all typically quite different so it's not at all clear which of the `config.h` files would be the right one to install.

> `WellContributions.hpp` should be able to be used in a CUDA-only environment and in an OpenCL-only environment. Would you propose a layer of inheritance?

You don't need any kind of inheritance.  You can continue to use the preprocessor symbols as usual.  You just need to ensure that every `.cpp` file that might end up including `WellContributions.hpp` brings in `config.h` as their first `#include` statement.